### PR TITLE
fix(erc4626): adapts slippage

### PR DESCRIPTION
**Pull request by Rubilmax**
- Fixes https://cantina.xyz/code/8409a0ce-6c21-4cc9-8ef2-bd77ce7425af/findings/b0413b48-5b50-4769-835a-b8ed1236b3bd
- Fixes https://cantina.xyz/code/8409a0ce-6c21-4cc9-8ef2-bd77ce7425af/findings/2bf73196-3629-4ed6-bd8c-3be1f1137bff
- Fixes https://cantina.xyz/code/8409a0ce-6c21-4cc9-8ef2-bd77ce7425af/findings/27b80ef4-c763-4769-9dde-515451c9f074
- Fixes https://cantina.xyz/code/8409a0ce-6c21-4cc9-8ef2-bd77ce7425af/findings/5cdb109a-ffd6-496a-a2bb-2c46eb768be1
- Fixes https://cantina.xyz/code/8409a0ce-6c21-4cc9-8ef2-bd77ce7425af/findings/65f8f161-f2ed-47dc-b6f2-d7fc6dccaf2c

### CODE REVIEW

**Comment by MerlinEgalite** [comment]
Tests are failing :/

**Review by adhusson — CHANGES_REQUESTED** [review]
I think tests for redeem, deposit and stake are missing.

**Review by adhusson — APPROVED** [review]

File: test/forge/ethereum/EthereumStEthBundlerEthereumTest.sol [hunk 1]
```diff
@@ -49,12 +49,29 @@ contract EthereumStEthBundlerEthereumTest is EthereumTest {
...
+        assertApproxEqAbs(IStEth(ST_ETH).sharesOf(USER), shares / 2, 2, "shares");
+    }
+
     function testStakeEthSlippageExceeded(uint256 amount) public {
         amount = bound(amount, MIN_AMOUNT, 10_000 ether);
 
         uint256 shares = IStEth(ST_ETH).getSharesByPooledEth(amount);
 
-        bundle.push(abi.encodeCall(StEthBundler.stakeEth, (amount, shares, address(0))));
+        bundle.push(abi.encodeCall(StEthBundler.stakeEth, (amount, shares - 2, address(0))));
```
Comment (MerlinEgalite) [inline test/forge/ethereum/EthereumStEthBundlerEthereumTest.sol:74]: - 1 does not work?

File: test/forge/ethereum/EthereumStEthBundlerEthereumTest.sol [hunk 2]
```diff
@@ -49,12 +49,29 @@ contract EthereumStEthBundlerEthereumTest is EthereumTest {
         assertApproxEqAbs(ERC20(ST_ETH).balanceOf(RECEIVER), amount, 3, "balanceOf(RECEIVER)");
     }
 
+    function testStakeEthSlippageAdapts(uint256 amount) public {
+        amount = bound(amount, MIN_AMOUNT, 10_000 ether);
+
+        uint256 shares = IStEth(ST_ETH).getSharesByPooledEth(amount);
+
+        bundle.push(abi.encodeCall(StEthBundler.stakeEth, (amount, shares - 2, address(0))));
```
Comment (MerlinEgalite) [inline test/forge/ethereum/EthereumStEthBundlerEthereumTest.sol:57]: same question as below

File: test/forge/ethereum/EthereumStEthBundlerEthereumTest.sol (hunk 1 above)
Comment (Rubilmax) [reply test/forge/ethereum/EthereumStEthBundlerEthereumTest.sol:74]: Nope, -1 is not enough

**Review by MerlinEgalite — APPROVED** [review]

File: src/ERC4626Bundler.sol [hunk 3]
```diff
@@ -22,8 +22,7 @@ abstract contract ERC4626Bundler is BaseBundler {
     /// @dev Initiator must have previously transferred their assets to the bundler.
     /// @dev Assumes the given `vault` implements EIP-4626.
     /// @param vault The address of the vault.
-    /// @param shares The amount of shares to mint. Pass `type(uint256).max` to mint the shares corresponding to the
-    /// bundler's assets.
+    /// @param shares The amount of shares to mint. Will be capped to the shares corresponding to the bundler's assets.
```
Comment (Jean-Grimal) [inline src/ERC4626Bundler.sol:25]:
```suggestion
    /// @param shares The amount of shares to mint. Will be capped at the shares corresponding to the bundler's assets.
```

File: src/ERC4626Bundler.sol [hunk 4]
```diff
@@ -34,23 +33,23 @@ abstract contract ERC4626Bundler is BaseBundler {
...
-        require(assets <= maxAssets, ErrorsLib.SLIPPAGE_EXCEEDED);
+        require(assets * initialShares <= maxAssets * shares, ErrorsLib.SLIPPAGE_EXCEEDED);
     }
 
     /// @notice Deposits the given amount of `assets` on the given ERC4626 `vault`, on behalf of `receiver`.
     /// @dev Initiator must have previously transferred their assets to the bundler.
     /// @dev Assumes the given `vault` implements EIP-4626.
     /// @param vault The address of the vault.
-    /// @param assets The amount of assets to deposit. Pass `type(uint256).max` to deposit the bundler's assets.
+    /// @param assets The amount of assets to deposit. Will be capped to the bundler's assets.
```
Comment (Jean-Grimal) [inline src/ERC4626Bundler.sol:52]:
```suggestion
    /// @param assets The amount of assets to deposit. Will be capped at the bundler's assets.
```

File: src/ERC4626Bundler.sol [hunk 5]
```diff
@@ -61,24 +60,24 @@ abstract contract ERC4626Bundler is BaseBundler {
...
+        require(shares * initialAssets >= minShares * assets, ErrorsLib.SLIPPAGE_EXCEEDED);
     }
 
     /// @notice Withdraws the given amount of `assets` from the given ERC4626 `vault`, transferring assets to
     /// `receiver`.
     /// @dev Assumes the given `vault` implements EIP-4626.
     /// @param vault The address of the vault.
-    /// @param assets The amount of assets to withdraw. Pass `type(uint256).max` to withdraw the assets corresponding to
-    /// the owner's shares.
+    /// @param assets The amount of assets to withdraw. Will be capped to the assets corresponding to the owner's
```
Comment (Jean-Grimal) [inline src/ERC4626Bundler.sol:79]:
```suggestion
    /// @param assets The amount of assets to withdraw. Will be capped at the assets corresponding to the owner's
```

File: src/ERC4626Bundler.sol [hunk 6]
```diff
@@ -93,18 +92,19 @@ abstract contract ERC4626Bundler is BaseBundler {
...
         uint256 shares = IERC4626(vault).withdraw(assets, receiver, owner);
-        require(shares <= maxShares, ErrorsLib.SLIPPAGE_EXCEEDED);
+        require(shares * initialAssets <= maxShares * assets, ErrorsLib.SLIPPAGE_EXCEEDED);
     }
 
     /// @notice Redeems the given amount of `shares` from the given ERC4626 `vault`, transferring assets to `receiver`.
     /// @dev Assumes the given `vault` implements EIP-4626.
     /// @param vault The address of the vault.
-    /// @param shares The amount of shares to burn. Pass `type(uint256).max` to redeem the bundler's shares.
+    /// @param shares The amount of shares to burn. Will be capped to the bundler's shares.
```
Comment (Jean-Grimal) [inline src/ERC4626Bundler.sol:107]:
```suggestion
    /// @param shares The amount of shares to burn. Will be capped at the bundler's shares.
```

File: src/StEthBundler.sol [hunk 7]
```diff
@@ -41,16 +41,17 @@ abstract contract StEthBundler is BaseBundler {
     /// @notice Stakes the given `amount` of ETH via Lido, using the `referral` id.
     /// @notice stETH tokens are received by the bundler and should be used afterwards.
     /// @dev Initiator must have previously transferred their ETH to the bundler.
-    /// @param amount The amount of ETH to stake. Pass `type(uint256).max` to stake all.
+    /// @param amount The amount of ETH to stake. Will be capped to the bundler's balance.
```
Comment (Jean-Grimal) [inline src/StEthBundler.sol:44]:
```suggestion
    /// @param amount The amount of ETH to stake. Will be capped at the bundler's balance.
```

**Review by Jean-Grimal — CHANGES_REQUESTED** [review]

**Review by Jean-Grimal — APPROVED** [review]

**Review by MerlinEgalite — APPROVED** [review]

**Comment by Rubilmax** [comment]
> I think that it's better to pass a share price as slippage parameter (simpler require)

But we need to define a large enough precision for all vault cases...

**Review by Jean-Grimal — APPROVED** [review]

**Comment by MathisGD** [comment]
blocked by #379

File: src/ERC4626Bundler.sol [hunk 8]
```diff
@@ -44,7 +44,7 @@ abstract contract ERC4626Bundler is BaseBundler {
     /// @dev Initiator must have previously transferred their assets to the bundler.
     /// @dev Assumes the given `vault` implements EIP-4626.
     /// @param vault The address of the vault.
-    /// @param assets The amount of assets to deposit. Pass `type(uint256).max` to deposit the bundler's assets.
+    /// @param assets The amount of assets to deposit. Will be capped at the bundler's assets.
```
Comment (QGarchery) [inline src/ERC4626Bundler.sol:47]: We use this comment everywhere for the same situation (using a min), I wouldn't change it just here.

Comment (QGarchery) [inline src/ERC4626Bundler.sol:47]: I would also add a warning, something like "beware that you can receive less shares than `minShares` if your balance is less than `assets`"

File: src/ERC4626Bundler.sol
Comment (QGarchery) [inline src/ERC4626Bundler.sol:1]: Do we want to do the same thing for the MorphoBundler ?

File: src/ERC4626Bundler.sol (hunk 8 above)
Comment (Rubilmax) [reply src/ERC4626Bundler.sol:47]: I agree this deserves a dedicated PR, but I do think we should distinguish between the `type(uint256).max` and the "always capped" case

File: src/ERC4626Bundler.sol
Comment (Rubilmax) [reply src/ERC4626Bundler.sol:1]: We don't perform `min` in MorphoBundler!

**Review by MerlinEgalite — APPROVED** [review]

**Review by Jean-Grimal — APPROVED** [review]

File: src/StEthBundler.sol [hunk 9]
```diff
@@ -41,16 +41,17 @@ abstract contract StEthBundler is BaseBundler {
     /// @notice Stakes the given `amount` of ETH via Lido, using the `referral` id.
     /// @notice stETH tokens are received by the bundler and should be used afterwards.
     /// @dev Initiator must have previously transferred their ETH to the bundler.
-    /// @param amount The amount of ETH to stake. Pass `type(uint256).max` to stake all.
+    /// @param amount The amount of ETH to stake. Will be capped at the bundler's balance.
```
Comment (QGarchery) [inline src/StEthBundler.sol:44]: I think you missed this spot to highlight the min behavior and revert the change on `type(uint256).max`

**Review by QGarchery — APPROVED** [review]

**Review by MerlinEgalite — APPROVED** [review]

File: src/StEthBundler.sol [hunk 10]
```diff
@@ -45,12 +45,13 @@ abstract contract StEthBundler is BaseBundler {
...
     /// @param referral The address of the referral regarding the Lido Rewards-Share Program.
     function stakeEth(uint256 amount, uint256 minShares, address referral) external payable protected {
+        uint256 initialAmount = amount;
         amount = Math.min(amount, address(this).balance);
 
         require(amount != 0, ErrorsLib.ZERO_AMOUNT);
 
         uint256 shares = IStEth(ST_ETH).submit{value: amount}(referral);
-        require(shares >= minShares, ErrorsLib.SLIPPAGE_EXCEEDED);
+        require(shares * initialAmount >= minShares * amount, ErrorsLib.SLIPPAGE_EXCEEDED);
```
Comment (QGarchery) [inline src/StEthBundler.sol:54]: Still missing that the minShares is scaled down to handle the min above

**Review by Jean-Grimal — APPROVED** [review]

**Review by MathisGD — APPROVED** [review]
There could be a comment // we verify that shares >= minShares * initialAssets/assets.

**Comment by StErMi** [comment]
I'll make an example for the `stakeEth` function because the logic is the same for all the other functions involved in the PR

So the user wants to exchange `1000 asset` for `1000 shares` for a `sharePrice` of 1
But in the contracts there are only `500 asset` at this point and the **actual** `sharePrice` is `0,9` (you get more shares)

`shares = 500 / 0,9 = ~555`

`shares * initialAmount >= minShares * amount` ⇒ `555 * 1000 >= 1000 * 500` ⇒ `TRUE`


So **this tx will not revert**, but the user **has minted fewer shares** compared to the `minShares` specified in the function

Is this the intended behavior? Because the number of shares minted is less than `minShares` and it contradicts the natspec/slippage protection input

**Comment by QGarchery** [comment]
I think your example fits the specification @StErMi, as it is written in the natspec that:
`This parameter is proportionally scaled down in case there are fewer assets than `assets` on the bundler.`
In effect, we are comparing the scaled down assets with the scaled down shares that are actually traded. This means that the slippage mechanism applies to share price, and protects against the negative impact of the share price variation.

Here the share price variation corresponds to positive slippage, so it's expected that this tx is accepted

**Pull request merged**